dbcon/mysql/ha_mcs_execplan.cpp

```diff
@@ -134,6 +135,146 @@ class RecursionCounter
+// the SELECT_LEX in a scenario where the select_handler execution has failed
+// and we want to fallback to the server execution (MCOL-4525). Here, we mimick
+// the creation of Item_cond_and using tableOnExprList and condList.
+void convertOuterJoinToInnerJoin(List<TABLE_LIST>* join_list,
+    std::tr1::unordered_map<TABLE_LIST*, std::vector<COND*>>& tableOnExprList,
+    std::vector<COND*>& condList,
+    std::tr1::unordered_map<TABLE_LIST*, uint>& tableOuterJoinMap)
+{
+    TABLE_LIST *table;
+    NESTED_JOIN *nested_join;
+    List_iterator<TABLE_LIST> li(*join_list);
+
+    while ((table = li++))
+    {
+        table_map used_tables;
+        table_map not_null_tables = (table_map) 0;
+
+        if ((nested_join = table->nested_join))
+        {
+            auto iter = tableOnExprList.find(table);
+
+            if ((iter != tableOnExprList.end()) && !iter->second.empty())
+            {
+                convertOuterJoinToInnerJoin(&nested_join->join_list,
+                    tableOnExprList, tableOnExprList[table], tableOuterJoinMap);
+            }
+
+            nested_join->used_tables = (table_map) 0;
+            nested_join->not_null_tables = (table_map) 0;
+
+            convertOuterJoinToInnerJoin(&nested_join->join_list,
+                tableOnExprList, condList, tableOuterJoinMap);
+
+            used_tables = nested_join->used_tables;
+            not_null_tables = nested_join->not_null_tables;
+        }
+        else
+        {
+            used_tables = table->get_map();
+
+            if (!condList.empty())
+            {
+                if (condList.size() == 1)
+                {
+                    not_null_tables = condList[0]->not_null_tables();
+                }
+                else
+                {
+                    calculateNotNullTables(condList, not_null_tables);
+                }
+            }
+        }
+
+        if (table->embedding)
+        {
+            table->embedding->nested_join->used_tables |= used_tables;
+            table->embedding->nested_join->not_null_tables |= not_null_tables;
+        }
+
+        if (!(table->outer_join & (JOIN_TYPE_LEFT | JOIN_TYPE_RIGHT)) ||
+            (used_tables & not_null_tables))
+        {
+            if (table->outer_join)
+            {
+                tableOuterJoinMap[table] = table->outer_join;
+            }
+
+            table->outer_join = 0;
+
+            auto iter = tableOnExprList.find(table);
+
+            if (iter != tableOnExprList.end() && !iter->second.empty())
+            {
+                // The original implementation in simplify_joins() creates
+                // an Item_cond_and object here. Instead of doing so, we
+                // append the table->on_expr to the condList and hence avoid
+                // making any permanent changes to SELECT_LEX::JOIN::conds.
+                condList.insert(condList.end(),
+                    tableOnExprList[table].begin(), tableOnExprList[table].end());
```

Comment:
IMHO this tabulation style is hard to read b/c I can't clearly see where the argument list begins.

Reply:
We have mixed use of this style in our code base:
```
some_func_name(some_var1,
    some_var2)
```
and style of the form:
```
some_func_name(some_var1,
               some_var2)
```
Not sure which one is more appropriate. I think use of one of these styles over the other would depend on how long the function name `some_func_name` is. Both should be OK depending on the context.

Reply:
Agree. I just personally prefer the 2nd style unless the initial indentation is too far from the left.